### DISCUSSION TITLE
Trouble running simulations on Pmetrics 2.0

### DISCUSSION THREAD

**Discussion by Leshtambe** [LAPKB/Pmetrics#195] [Q&A]
Hello there,

I am having issues running sims in the new version of Pmetrics. 

When I write this command: PM_sim$run(poppar=final.398, model="lalala.txt", data="dododo.csv", limits=NA, nsim=1000, predInt=c(144, 168, 0.5), covariate=covariate)

This error comes up: Error in SIMrun(poppar = final.398, model = "vancA4.txt", data = "GanWtsim.csv",  : 
  Ensure poppar is a list of 3 elements: weight, mean, covariance.

Can anyone help, please?

**Comment by mnneely** [comment]
Thanks for posting here. It looks like you’ve got a mixture of old and new Pmetrics going on.

Try something like this. This code should still work on output from a run made under old Pmetrics (prior to version 2).

```
run398 <- PM_load(398) #note the underscore
sim_data <- PM_data$new(“GatWtsim.csv”) #assumes this file in current working directory
sim398 <- run398$sim(data = sim_data, limits = NA) 
```
With this approach, you activate the simulation method attached to the run398 PM_result object. That object contains poppar as run39$final, which is the equivalent of final.398 in old Pmetrics. It also contains the model as run398$model, and the data as run398$data. So when you simulate this way, you only need to supply objects that are different than the ones in the PM_result object, typically the data. Hence, the code above only has a data argument, under the assumption that “GanWtsim.csv” is different than the data file you used to fit the model. You could also substitute a different model, assuming the parameters were the same as in run398$final.

If you don’t have a prior run, say when simulating from the literature, then you use code like this.
```
my_poppar <- list(wt = 1, mean = c(…), cov = matrix(…))
sim_data <- PM_data$new(“GatWtsim.csv”)
sim_model <- PM_model$new(“vancA4.txt”)
sim1398<- PM_sim$new(poppar = my_poppar, model = sim_model, data = sim_data, limits = NA) 
```

Note in the second approach, we are calling the simulator from scratch rather than activating the simulation method on a PM_result object as we do in the first approach. So in the second approach, we have to supply the poppar, model, and data objects, while in the first approach, we only supply what is different than the objects contained in the PM_result object made when we used PM_load.

While it may seem more steps to first make data or model objects if needed with PM_data$new() or PM_model$new(), the advantage is that they are now in memory in R and you no longer have to move them around in your file system to ensure that the correct file is present. You can also plot the model now, giving you a visual representation to ensure it’s what you want: `sim_model$plot()`.

If you are using the results of a prior NPAG or IT2B run to simulate, use the first approach always. Pmetrics will find the $final, $data, and $model objects and use them for the simulation, unless you override the data or model as above. The second approach is really only useful for de novo simulations with manually created poppar objects when simulating without a prior NPAG/IT2B run.

Have a look at the [Simulation](https://lapkb.github.io/Pmetrics/articles/simulation.html) article in the help documents.

Hope that makes it clearer.
